### PR TITLE
[2/5] light client http api

### PR DESCRIPTION
**Pull request by nicopernas**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the Altair spec logic defined in this version:

https://github.com/ethereum/consensus-specs/tree/208da34ac4e75337baf79adebf036ab595e39f15/specs/altair/light-client

It also implements the relevant endpoints specified at https://github.com/ethereum/beacon-APIs

https://github.com/prysmaticlabs/prysm/pull/11815

**Which issues(s) does this PR fix?**

Partially addresses https://github.com/prysmaticlabs/prysm/issues/11571

**Other notes for review**

This PR is a follow up to https://github.com/prysmaticlabs/prysm/pull/11815 which was too big to be reviewed. This is the second PR in a series of 5 that I will be submitting sequentially. 

The heavy lifting here was done by [@qinlz2](https://github.com/qinlz2) and [@7AC](https://github.com/7AC), so kudos to them. 

-------
Full series:

- [1/5] https://github.com/prysmaticlabs/prysm/pull/12853
- [2/5] https://github.com/prysmaticlabs/prysm/pull/12984 (this PR)
- [3/5]
- [4/5]
- [5/5] https://github.com/prysmaticlabs/prysm/pull/12854
-----
https://github.com/prysmaticlabs/prysm/pull/12854 is the last PR in the series that contains **all** changes so you can go back and see the full picture while reviewing this one.

### CODE REVIEW

**Comment by rkapka** [comment]
Hi @nicopernas , can you please open one more PR and move `lightclient.go` from the blockchain package to `core/light-client`? The argument is:

> they are all stateless methods with no requirement to use the blockchain service

**Comment by nicopernas** [comment]
> Hi @nicopernas , can you please open one more PR and move `lightclient.go` from the blockchain package to `core/light-client`? The argument is:
> 
> 
> 
> > they are all stateless methods with no requirement to use the blockchain service

That's no problem at all. If it's not an urgent change, can we do it after the rest of the PRs are merged?? It would save me from rebasing once or twice :)

**Comment by rkapka** [comment]
OK, but in that case I will open an issue for it, just to make sure this is not forgotten.

**Comment by nicopernas** [comment]
> OK, but in that case I will open an issue for it, just to make sure this is not forgotten.

Thanks for that!

File: beacon-chain/rpc/eth/light-client/handlers_test.go [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+	request = mux.SetURLVars(request, muxVars)
+	writer := httptest.NewRecorder()
+	writer.Body = &bytes.Buffer{}
+
+	s.GetLightClientBootstrap(writer, request)
+	require.Equal(t, http.StatusOK, writer.Code)
+	resp := &LightClientBootstrapResponse{}
+	require.NoError(t, json.Unmarshal(writer.Body.Bytes(), resp))
+	require.Equal(t, "CAPELLA", resp.Version)
+	require.Equal(t, hexutil.Encode(header.Header.BodyRoot), resp.Data.Header.BodyRoot)
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers_test.go:80]: Given that we have the helper function that prepares the whole bootstrap object, I think it's fine to check only the version here and maybe verify that `resp.Data` is not nil. But I would still check values of several bootstrap fields by writing a unit test of the helper function.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	stateRoot, err := state.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not get state root %v", err)
+	}
+
+	// Return result
+	result := &LightClientBootstrap{
+		Header: &apimiddleware.BeaconBlockHeaderJson{
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:86]: Similarly comment as for `apimiddleware.SyncCommitteeJson`, but this time please put the new type inside `/eth/shared/structs_blocks.go`

File: beacon-chain/rpc/eth/light-client/structs.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+type LightClientBootstrapResponse struct {
+	Version string                `json:"version"`
+	Data    *LightClientBootstrap `json:"data"`
+}
+
+type LightClientBootstrap struct {
+	Header                     *apimiddleware.BeaconBlockHeaderJson `json:"header"`
+	CurrentSyncCommittee       *apimiddleware.SyncCommitteeJson     `json:"current_sync_committee"`
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/structs.go:14]: these two fields should be of the newly created types that I mentioned in other comments, not the ones in `apimiddleware`

**Review by rkapka — CHANGES_REQUESTED** [review]

**Comment by rkapka** [comment]
I reviewed the bootstrap function and will likely continue tomorrow.

**Comment by rkapka** [comment]
I found something missing in your previous PR. In your proto definitions you use `current_sync_committee_branch.depth`, `next_sync_committee_branch.depth` and `finality_branch.depth`, but in order for these variables to work with SSZ, you need to define their values in `proto/ssz_proto_library.bzl`. Will this be done in one of the next PRs? Because if not and this is an omission, then I don't understand how SSZ generation for these proto messages could ever work.

Looking at these variables, will `finality_branch.depth` differ between minimal and mainnet? If not, we can simply hardcode it as a number.

Speaking about SSZ generation, API spec requires SSZ support:
> Depending on the Accept header it can be returned either as JSON or SSZ-serialized bytes.

This would mean that SSZ generation for `LightClientBootstrap` and other types is indeed necessary. Which brings a question: do you have SSZ support in APIs coming in a later PR? If not, we can create an issue for it.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate next sync committee and proof
+	var nextSyncCommittee *v2.SyncCommittee
+	var nextSyncCommitteeBranch [][]byte
+
+	// update_signature_period = compute_sync_committee_period(compute_epoch_at_slot(block.message.slot))
+	updateSignaturePeriod := uint64(block.Block().Slot()) / slotsPerPeriod
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:175]: you can use the `slots.ToEpoch` helper

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+	// update_attested_period = compute_sync_committee_period(compute_epoch_at_slot(attested_header.slot))
+	updateAttestedPeriod := uint64(result.AttestedHeader.Slot) / slotsPerPeriod
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:178]: same here

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+		nextSyncCommittee = &v2.SyncCommittee{
+			Pubkeys:         tempNextSyncCommittee.Pubkeys,
+			AggregatePubkey: tempNextSyncCommittee.AggregatePubkey,
+		}
+
+		nextSyncCommitteeBranch, err = attestedState.NextSyncCommitteeProof(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("could not get next sync committee proof %v", err)
+		}
+	} else {
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/helpers.go:195]: Is there any gain in returning these dummy slices instead of keeping this fields nil?

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+	slotsPerPeriod := uint64(config.EpochsPerSyncCommitteePeriod) * uint64(config.SlotsPerEpoch)
+
+	// Adjust count based on configuration
+	countParam := req.URL.Query().Get("count")
+	count, err := strconv.ParseUint(countParam, 10, 64)
+	if err != nil {
+		http2.HandleError(w, fmt.Sprintf("got invalid 'count' query variable '%s', err %v", countParam, err),
+			http.StatusInternalServerError)
+		return
+	}
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers.go:81]: there is a `shared.UintFromQuery` helper that you can use

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+
+	// Determine the start and end periods
+	startPeriodParam := req.URL.Query().Get("start_period")
+	startPeriod, err := strconv.ParseUint(startPeriodParam, 10, 64)
+	if err != nil {
+		http2.HandleError(w, fmt.Sprintf("got invalid 'start_period' query variable '%s', err %v", startPeriodParam, err),
+			http.StatusInternalServerError)
+		return
+	}
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers.go:90]: same here

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+		startPeriod = uint64(config.AltairForkEpoch) * uint64(config.SlotsPerEpoch) / slotsPerPeriod
+	}
+
+	headState, err := bs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		http2.HandleError(w, "could not get head state: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	lHeadSlot := uint64(headState.Slot())
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers.go:109]: Can this be renamed to `headSlot`? We don't use such prefixes anywhere in the codebase. Similarly for other variables in this function

**Comment by nicopernas** [comment]
Hey @rkapka thanks a lot for all the good feedback! We are making internal changes to our team structure and I will be looking into something else from now own. Someone else from the team will take over to make sure this and future PRs make it to `develop`.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+	}
+
+	block, err := bs.BeaconDB.Block(ctx, latestBlockHeaderRoot)
+	if err != nil {
+		return nil, fmt.Errorf("could not get latest block %v", err)
+	}
+	if block == nil {
+		return nil, fmt.Errorf("latest block is nil")
+	}
+	// Loop through the blocks until we find a block that has super majority of sync committee signatures (2/3)
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers.go:377]: Are you sure this comment is correct? Isn't supermajority required only when calling from `GetLightClientFinalityUpdate`?

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+	response := &LightClientUpdateWithVersion{
+		Version: ethpbv2.Version(attestedState.Version()).String(),
+		Data:    update,
+	}
+
+	http2.WriteJson(w, response)
+}
+
+// getLightClientEventBlock - returns the block that should be used for light client events, which satisfies the minimum number of signatures from sync committee
+func (bs *Server) getLightClientEventBlock(ctx context.Context, minSignaturesRequired uint64) (interfaces.ReadOnlySignedBeaconBlock, error) {
```
Comment (rkapka) [inline beacon-chain/rpc/eth/light-client/handlers.go:351]: Please add some unit tests, two at a minimum:
- happy case
- number of signatures too low, need to fetch parent block

**Review by rkapka — CHANGES_REQUESTED** [review]

**Comment by qinlz2** [comment]
@rkapka This is `Lizhang Qin` from `polymer labs`.   I will take over this PR work from @nicopernas . 

I will address your comments and submit revisions.  This will take some time.    Thanks!

**Comment by rkapka** [comment]
Hi @qinlz2 , are you able to say how when you might have some time to pick this up? We would like to give light client support higher priority and implement this in the next few months.

**Comment by qinlz2** [comment]
Yes, commit hours for this now.   Was deviate from this PR (and the rest of PRs) a little bit.  
Coming back now, we will try to meet your timeline.   Thanks!

> Hi @qinlz2 , are you able to say how when you might have some time to pick this up? We would like to give light client support higher priority and implement this in the next few months (by "this" I mean not this one PR, but the whole thing).

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+// GetLightClientBootstrap - implements https://github.com/ethereum/beacon-APIs/blob/263f4ed6c263c967f13279c7a9f5629b51c5fc55/apis/beacon/light_client/bootstrap.yaml
+func (bs *Server) GetLightClientBootstrap(w http.ResponseWriter, req *http.Request) {
+	// Prepare
+	ctx, span := trace.StartSpan(req.Context(), "beacon.GetLightClientBootstrap")
+	defer span.End()
+
+	// Get the block
+	blockRootParam, err := hexutil.Decode(mux.Vars(req)["block_root"])
+	if err != nil {
+		http2.HandleError(w, "invalid block root "+err.Error(), http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:32]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+		return
+	}
+
+	var blockRoot [32]byte
+	copy(blockRoot[:], blockRootParam)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:37]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+	}
+
+	bootstrap, err := CreateLightClientBootstrap(ctx, state)
+	if err != nil {
+		http2.HandleError(w, "could not get light client bootstrap "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	response := &LightClientBootstrapResponse{
+		Version: ethpbv2.Version(blk.Version()).String(),
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:57]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+//	        state_root=hash_tree_root(state),
+//	        body_root=state.latest_block_header.body_root,
+//	    ),
+//	    current_sync_committee=state.current_sync_committee,
+//	    current_sync_committee_branch=compute_merkle_proof_for_state(state, CURRENT_SYNC_COMMITTEE_INDEX)
+//	)
+func CreateLightClientBootstrap(ctx context.Context, state state.BeaconState) (*LightClientBootstrap, error) {
+	// assert compute_epoch_at_slot(state.slot) >= ALTAIR_FORK_EPOCH
+	if slots.ToEpoch(state.Slot()) < params.BeaconConfig().AltairForkEpoch {
+		return nil, fmt.Errorf("invalid state slot %d", state.Slot())
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:45]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	// assert state.slot == state.latest_block_header.slot
+	if state.Slot() != state.LatestBlockHeader().Slot {
+		return nil, fmt.Errorf("invalid state slot %d", state.Slot())
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:50]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	// Prepare data
+	latestBlockHeader := state.LatestBlockHeader()
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:54]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+
+	currentSyncCommittee, err := state.CurrentSyncCommittee()
+	if err != nil {
+		return nil, fmt.Errorf("could not get current sync committee %v", err)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:58]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	currentSyncCommitteePubkeys := currentSyncCommittee.GetPubkeys()
+	committee := apimiddleware.SyncCommitteeJson{
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:62]: @rkapka We have multiple places using `apimiddleware`, such as using `apimiddleware.BeaconBlockHeaderJson`. 

Should we also introduce shared structs for `BeaconBlockHeader`?

[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:62]: No worry, I see the same comment below.  We will handle it.

[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:62]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+	}
+	for i, pubkey := range currentSyncCommitteePubkeys {
+		committee.Pubkeys[i] = hexutil.Encode(pubkey)
+	}
+	currentSyncCommitteeProof, err := state.CurrentSyncCommitteeProof(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not get current sync committee proof %v", err)
+	}
+
+	branch := make([]string, len(currentSyncCommitteeProof))
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:74]: Yes, this is true.  Normally, it is safer to use len (which is cheap as well) to avoid possible mem issue. 
But in this special case, we can use constant.  

Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	v1 "github.com/prysmaticlabs/prysm/v4/proto/eth/v1"
+	v2 "github.com/prysmaticlabs/prysm/v4/proto/eth/v2"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+)
+
+const (
+	nextSyncCommitteeBranchNumOfLeaves = 5
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:23]: This `const` may be wrongly named.   It basically states the number of branches needed for a proof. 

This is only relevant the position of the field in the trie leaves.  (Where currentSyncCommittee and nextSyncCommittee are next to each  other). 

So this isn't related to `sync committee` size (no matter it is 32 or 512), but related to the total number of leaves in the trie. 

Yes making this configurable is future proof though.   So let's make it configurable.

[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:23]: Patched.

File: beacon-chain/rpc/eth/light-client/server.go [hunk 22]
```diff
@@ -0,0 +1,13 @@
+package lightclient
+
+import (
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/rpc/lookup"
+)
+
+type Server struct {
+	BeaconDB    db.ReadOnlyDatabase
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/server.go:10]: Good to know.  Thanks. 

Patched.

File: beacon-chain/rpc/eth/light-client/handlers_test.go (hunk 1 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers_test.go:80]: At this moment, let's just check `resp.Data` not nil, and existing very basic field  checking. 
Feel free to create a following task, so we can add more solid unit tests.

File: beacon-chain/rpc/eth/light-client/helpers.go (hunk 2 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:86]: Patched.

File: beacon-chain/rpc/eth/light-client/structs.go (hunk 3 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/structs.go:14]: Patched.

**Comment by qinlz2** [comment]
> current_sync_committee_branch.depth

@rkapka In our initial PRs, we only implements JSON support. 
For this PRs series we are trying to merge, let's just do JSON part.  

Please create task for later SSZ support.  Thanks!

**Comment by qinlz2** [comment]
> I found something missing in your previous PR. In your proto definitions you use `current_sync_committee_branch.depth`, `next_sync_committee_branch.depth` and `finality_branch.depth`, but in order for these variables to work with SSZ, you need to define their values in `proto/ssz_proto_library.bzl`. Will this be done in one of the next PRs? Because if not and this is an omission, then I don't understand how SSZ generation for these proto messages could ever work.
> 
> Looking at these variables, will `finality_branch.depth` differ between minimal and mainnet? If not, we can simply hardcode it as a number.
> 
> Speaking about SSZ generation, API spec requires SSZ support:
> 
> > Depending on the Accept header it can be returned either as JSON or SSZ-serialized bytes.
> 
> This would mean that SSZ generation for `LightClientBootstrap` and other types is indeed necessary. Which brings a question: do you have SSZ support in APIs coming in a later PR? If not, we can create an issue for it.

Between mainnet and min, there is no difference.  We can hardcode (and we actually didn't use these at all in the code I think).   So I will remove them for now from proto definitions.

File: beacon-chain/rpc/eth/light-client/helpers.go (hunk 4 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:166]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go (hunk 5 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:169]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go (hunk 6 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:186]: This is just the spec logic. 

`
//	else:
//	    next_sync_committee = SyncCommittee()
//	    next_sync_committee_branch = [Bytes32() for _ in range(floorlog2(NEXT_SYNC_COMMITTEE_INDEX))]
`

For now, should we just keep it?

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 7 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 8 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 9 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:109]: Sure, patched.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+		var state state.BeaconState
+		var block interfaces.ReadOnlySignedBeaconBlock
+		for lSlot := lLastSlotInPeriod; lSlot >= lFirstSlotInPeriod; lSlot-- {
+			state, err = bs.Stater.StateBySlot(ctx, types.Slot(lSlot))
+			if err != nil {
+				continue
+			}
+
+			// Get the block
+			latestBlockHeader := *state.LatestBlockHeader()
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:139]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 24]
```diff
@@ -0,0 +1,402 @@
...
+
+		update, err := CreateLightClientUpdate(
+			ctx,
+			slotsPerPeriod,
+			state,
+			block,
+			attestedState,
+			finalizedBlock,
+		)
+
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:205]: Patched.

File: beacon-chain/rpc/eth/light-client/helpers.go [hunk 25]
```diff
@@ -0,0 +1,377 @@
...
+//	return LightClientUpdate(
+//	    attested_header=attested_header,
+//	    next_sync_committee=next_sync_committee,
+//	    next_sync_committee_branch=next_sync_committee_branch,
+//	    finalized_header=finalized_header,
+//	    finality_branch=finality_branch,
+//	    sync_aggregate=block.message.body.sync_aggregate,
+//	    signature_slot=block.message.slot,
+//	)
+func CreateLightClientUpdate(
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:149]: You are correct.   There is no implementation for above mentioned text section.  
We only implemented the python pseudo spec. 

Can we create a separate task to patch this?  As this function is individual function, not used by others.

File: beacon-chain/rpc/eth/light-client/helpers.go
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/helpers.go:1]: Yes, you are correct. 

Patched, only `NewLightClientBootstrapFromJSON` is exported, and this is used in light client side.

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 10 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:377]: Good catch, the comment isn't correct. 

Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:346]: We didn't consider the cache.  Please add a task, so we can improve this later. 

Depending on the configuration, typically, satisfying the loop condition isn't hard.  (But yes, in extreme error conditions, loop will be triggered severely, and causing performance issue). 

BTW, since we switch to use `lookup.Blocker`, does this component provide cache automatically?  If so, then we don't need to add cache.

File: beacon-chain/rpc/eth/light-client/handlers.go [hunk 26]
```diff
@@ -0,0 +1,402 @@
...
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+	types "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	http2 "github.com/prysmaticlabs/prysm/v4/network/http"
+	ethpbv2 "github.com/prysmaticlabs/prysm/v4/proto/eth/v2"
+	"github.com/wealdtech/go-bytesutil"
+	"go.opencensus.io/trace"
+)
+
+// GetLightClientBootstrap - implements https://github.com/ethereum/beacon-APIs/blob/263f4ed6c263c967f13279c7a9f5629b51c5fc55/apis/beacon/light_client/bootstrap.yaml
+func (bs *Server) GetLightClientBootstrap(w http.ResponseWriter, req *http.Request) {
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go:24]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers_test.go [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+func TestLightClientHandler_GetLightClientUpdatesByRange(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers_test.go:80]: Patched.

File: beacon-chain/rpc/eth/light-client/handlers.go (hunk 11 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/light-client/handlers.go]: Patched.

**Comment by qinlz2** [comment]
@rkapka I pushed new revision based on your valuable comments.  
Before I rebase to the latest baseline, please help to take a look first, and we will address your comments/advice. 
Thanks!

**Comment by rkapka** [comment]
Hey @qinlz2 , everything looks great! My only concern is https://github.com/prysmaticlabs/prysm/pull/12984#discussion_r1350344110 - test names don't align with the test scenarios that I listed. Can you confirm that all test cases are covered? Maybe test names are incorrect?

**Comment by rkapka** [comment]
I updated https://github.com/prysmaticlabs/prysm/issues/12991 to serve as a general tracking issue for future light client improvements.

File: beacon-chain/rpc/eth/shared/structs.go [hunk 28]
```diff
@@ -536,6 +542,41 @@ func (m *SyncCommitteeMessage) ToConsensus() (*eth.SyncCommitteeMessage, error)
 	}, nil
 }
 
+func SyncCommitteeFromConsensus(sc *eth.SyncCommittee) *SyncCommittee {
+	if sc == nil {
+		return nil
+	}
```
Comment (rkapka) [inline beacon-chain/rpc/eth/shared/structs.go:548]: Can you remove this?

**Comment by qinlz2** [comment]
> Hey @qinlz2 , everything looks great! My only concern is [#12984 (comment)](https://github.com/prysmaticlabs/prysm/pull/12984#discussion_r1350344110) - test names don't align with the test scenarios that I listed. Can you confirm that all test cases are covered? Maybe test names are incorrect?

Yes, all of them are covered and tested.   (Due to slight code sequence re-order, the naming for test case won't be exactly the same as your original comment, but all the important `if-else` branches are tested)

File: beacon-chain/rpc/eth/shared/structs.go (hunk 28 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/shared/structs.go:548]: Sure, patched.

**Comment by qinlz2** [comment]
@rkapka I update slightly based on your recent comment.  
Do you need me to rebase first?

**Comment by rkapka** [comment]
> > Hey @qinlz2 , everything looks great! My only concern is [#12984 (comment)](https://github.com/prysmaticlabs/prysm/pull/12984#discussion_r1350344110) - test names don't align with the test scenarios that I listed. Can you confirm that all test cases are covered? Maybe test names are incorrect?
> 
> Yes, all of them are covered and tested. (Due to slight code sequence re-order, the naming for test case won't be exactly the same as your original comment, but all the important `if-else` branches are tested)

Perfect! In that case let's fix conflicts and merge this beauty :-)

**Comment by qinlz2** [comment]
> > > Hey @qinlz2 , everything looks great! My only concern is [#12984 (comment)](https://github.com/prysmaticlabs/prysm/pull/12984#discussion_r1350344110) - test names don't align with the test scenarios that I listed. Can you confirm that all test cases are covered? Maybe test names are incorrect?
> > 
> > 
> > Yes, all of them are covered and tested. (Due to slight code sequence re-order, the naming for test case won't be exactly the same as your original comment, but all the important `if-else` branches are tested)
> 
> @qinlz2 Perfect! In that case let's fix conflicts and merge this beauty :-)

Thanks, @rkapka  
I rebased to latest `develop` branch, manually tested it.  

It should be good to go now.

**Comment by rkapka** [comment]
Hey @qinlz2 , please run `bazel run //:gazelle -- fix`

**Comment by rkapka** [comment]
Also `TestGetSpec` fails:
```
assertions.go:37: handlers_test.go:174 Values are not equal, want: 112 (int), got: 113 (int)
handlers_test.go:419: Incorrect key: MAX_REQUEST_LIGHT_CLIENT_UPDATES
```

**Comment by qinlz2** [comment]
> Also `TestGetSpec` fails:
> 
> ```
> assertions.go:37: handlers_test.go:174 Values are not equal, want: 112 (int), got: 113 (int)
> handlers_test.go:419: Incorrect key: MAX_REQUEST_LIGHT_CLIENT_UPDATES
> ```

@rkapka Thanks, patched both.

**Review by rkapka — APPROVED** [review]
